File: hackerrank/problem solving/algorithms/1 - warmup/6 - plus minus/plus_minus.cpp
// https://www.hackerrank.com/challenges/plus-minus/problem?isFullScreen=true
// C++23

import std;
using namespace std;

vector<int> read_numbers(int n);
vector<double> plus_minus(const vector<int>& arr);
void print_ratios(const vector<double>& ratios);

int main()
{
    int n;
    cin >> n;
    vector arr{ read_numbers(n) };
    vector ratios{ plus_minus(arr) };
    print_ratios(ratios);
    return 0;
}

vector<int> read_numbers(int n)
{
    vector<int> arr(n);
    for (auto& x : arr)
        cin >> x;
    return arr;
}

// n: length of arr
// T: O(n)
// S: O(1) extra space
vector<double> plus_minus(const vector<int>& arr)
{
    int positive{}, negative{}, zero{};

    for (int number : arr)
        if (number > 0)
            ++positive;
        else if (number < 0)
            ++negative;
        else
            ++zero;

    double n{ static_cast<double>(arr.size()) };
    return { positive / n, negative / n, zero / n };
}

void print_ratios(const vector<double>& ratios)
{
    cout << fixed << setprecision(6);
    for (auto ratio : ratios)
        cout << ratio << '\n';
}

File: hackerrank/problem solving/algorithms/1 - warmup/6 - plus minus/plus_minus.rs
// https://www.hackerrank.com/challenges/plus-minus/problem?isFullScreen=true

use text_io::read;

fn main() {
    let n: usize = read!();
    let arr: Vec<i32> = read_numbers(n);
    let ratios: Vec<f32> = plus_minus(&arr);
    print_ratios(ratios);
}

fn read_numbers(n: usize) -> Vec<i32> {
    (0..n).map(|_| read!()).collect()
}

// n: length of arr
// T: O(n)
// S: O(1) extra space
fn plus_minus(arr: &[i32]) -> Vec<f32> {
    let mut positive: i32 = 0;
    let mut negative: i32 = 0;
    let mut zero: i32 = 0;

    for number in arr.iter() {
        if *number > 0 {
            positive += 1;
        } else if *number < 0 {
            negative += 1;
        } else {
            zero += 1;
        }
    }

    let n = arr.len() as f32;
    vec![positive as f32 / n, negative as f32 / n, zero as f32 / n]
}

fn print_ratios(ratios: Vec<f32>) {
    for ratio in ratios.iter() {
        println!("{:.6}", ratio);
    }
}
